### PR TITLE
CPM-493: add user category permission validation

### PR DESCRIPTION
**Pull request by nmarniesse**
- Validate the user exists
- Validate the user owns the product (through product categories)

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Symfony/Resources/config/validation/product_commands.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Akeneo\Pim\Enrichment\Product\Api\Command\UpsertProductCommand:
+    group_sequence:
+        - User
+        - UpsertProductCommand
+    properties:
+        userId:
+            - Akeneo\Pim\Enrichment\Product\Infrastructure\Validation\UserShouldExist:
+                  groups: [User]
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Symfony/Resources/config/validation/product_commands.yml:8]: The user is checked first. If it does not exist, the validation stops thanks to the group_sequence. It allows to:
- not perform other validations
- symplify the other validations (we don't have to handle the case where the user exists or not)

File: tests/back/Integration/IntegrationTestsBundle/Security/SystemUserAuthenticator.php [hunk 2]
```diff
@@ -58,6 +50,8 @@ public function createSystemUser()
             $user->addRole($role);
         }
 
+        $this->userSaver->save($user);
```
Comment (nmarniesse) [inline tests/back/Integration/IntegrationTestsBundle/Security/SystemUserAuthenticator.php:53]: In EE we add a query that needs the user id. So we have to create the user in DB to have an id and to make the query works.

File: src/Akeneo/Pim/Enrichment/Product/back/Api/Command/UpsertProductCommand.php [hunk 3]
```diff
@@ -27,7 +27,6 @@ public function __construct(
         private $associationsUserIntent = null,
         private array $valuesUserIntent = []
     ) {
-        Assert::stringNotEmpty($this->productIdentifier);
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Api/Command/UpsertProductCommand.php:30]: It's now a part of the validation process

File: src/Akeneo/Pim/Enrichment/Product/back/Domain/Query/IsUserCategoryGranted.php [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+use Akeneo\Pim\Enrichment\Product\Domain\Model\ProductIdentifier;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface IsUserCategoryGranted
+{
+    public function forProductAndAccessLevel(
```
[the remark this answers is not in the document]
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/Domain/Query/IsUserCategoryGranted.php:15]: Move in Category, returns own categories from a list of categories

[the remark this answers is not in the document]
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/Domain/Query/IsUserCategoryGranted.php:15]: https://github.com/akeneo/pim-enterprise-dev/blob/master/src/Akeneo/Pim/Permission/Bundle/Persistence/Sql/GetViewableCategoryCodesForUser.php

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/AntiCorruptionLayer/Feature.php [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+use Akeneo\Platform\Bundle\FeatureFlagBundle\FeatureFlags;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class Feature
+{
+    public const PERMISSION = 'permission';
```
[the remark this answers is not in the document]
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/AntiCorruptionLayer/Feature.php:15]: move in platform / feature flag

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+use PHPUnit\Framework\Assert;
+
+final class UpsertProductWithPermissionIntegration extends TestCase
+{
+    private UpsertProductHandler $upsertProductHandler;
+    private ProductRepositoryInterface $productRepository;
+
+    protected function getConfiguration(): Configuration
+    {
+        return $this->catalog->useTechnicalCatalog();
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php:33]: use minimal instead

File: src/Akeneo/Pim/Enrichment/Category/back/Domain/Query/DumbGetOwnedCategories.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Akeneo\Pim\Enrichment\Category\Domain\Query;
+
+use Akeneo\Pim\Enrichment\Category\API\Query\GetOwnedCategories;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class DumbGetOwnedCategories implements GetOwnedCategories
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Category/back/Domain/Query/DumbGetOwnedCategories.php:13]:
```suggestion
final class DummyGetOwnedCategories implements GetOwnedCategories
```

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    /** @test */
+    public function it_throws_an_exception_when_user_category_is_not_granted(): void
+    {
+        // Creates empty product (use command/handler when we can set a category)
+        $product = $this->get('pim_catalog.builder.product')->createProduct('identifier');
+        $this->get('pim_catalog.updater.product')->update($product, [
+            'categories' => ['print'],
+        ]);
+        $this->get('pim_catalog.saver.product')->save($product);
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php:52]: use createProduct()

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Query/SqlGetCategoryCodesIntegration.php [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            $this->sqlGetCategoryCodes->fromProductIdentifiers([ProductIdentifier::fromString('product_with_categories')])
+        );
+        Assert::assertEqualsCanonicalizing(
+            [
+                'product_without_category' => [],
+                'product_with_categories' => ['suppliers', 'print'],
+            ],
+            $this->sqlGetCategoryCodes->fromProductIdentifiers([
+                ProductIdentifier::fromString('product_without_category'),
+                ProductIdentifier::fromString('product_with_categories'),
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Query/SqlGetCategoryCodesIntegration.php:56]: Add test for unknown product

**Review by pierallard — APPROVED** [review]

**Review by cemorin — APPROVED** [review]

**Pull request merged**